### ISSUE TITLE
Sigma 18-35mm f/1.8 DC HSM misdetected after upgrading to firmware 2.03

### ISSUE THREAD

**Issue by lbschenkel** [Exiv2/exiv2#1368]
**Describe the bug**
Sigma 18-35mm f/1.8 DC HSM on a Canon 70D body, after upgrading to firmware 2.03 via the Sigma USB mount, is misdetected. Apparently the new firmware changed `LensType` from `150` to `368`.

Firmware changelog: https://www.sigma-global.com/en/download/lenses/firmware/popups/a_18_35_18/
Analogous report to `exiftool`: https://exiftool.org/forum/index.php?topic=3833.msg63023#msg63023

**To Reproduce**
Steps to reproduce the behaviour:

1. Upgrade lens to latest firmware (2.03)
2. `exifv2 --version`: `0.27.3`
3. Lens has a new type and is misdetected (two RAW files, same lens and body, old and new firmware):
```
$ exiv2 -p v 20200921_1828.cr2 20201017_2113.cr2
20200921_1828.cr2  0x0016 CanonCs      LensType                    Short       1  150
20200921_1828.cr2  0x0017 CanonCs      Lens                        Short       3  35 18 1
20200921_1828.cr2  0x0095 Canon        LensModel                   Ascii      74  18-35mm
20200921_1828.cr2  0xa432 Photo        LensSpecification           Rational    4  18/1 35/1 0/1 0/1
20200921_1828.cr2  0xa434 Photo        LensModel                   Ascii       8  18-35mm
20200921_1828.cr2  0xa435 Photo        LensSerialNumber            Ascii      11  0000000000
20201017_2113.cr2  0x0016 CanonCs      LensType                    Short       1  368
20201017_2113.cr2  0x0017 CanonCs      Lens                        Short       3  35 18 1
20201017_2113.cr2  0x0095 Canon        LensModel                   Ascii      74  18-35mm F1.8 DC HSM | Art 013
20201017_2113.cr2  0xa432 Photo        LensSpecification           Rational    4  18/1 35/1 0/1 0/1
20201017_2113.cr2  0xa434 Photo        LensModel                   Ascii      30  18-35mm F1.8 DC HSM | Art 013
20201017_2113.cr2  0xa435 Photo        LensSerialNumber            Ascii      11  0000000000

$ exiv2 -p a 20200921_1828.cr2 20201017_2113.cr2
20200921_1828.cr2  Exif.CanonCs.LensType                        Short       1  Sigma 18-35mm f/1.8 DC HSM | A
20200921_1828.cr2  Exif.CanonCs.Lens                            Short       3  18.0 - 35.0 mm
20200921_1828.cr2  Exif.Canon.LensModel                         Ascii      74  18-35mm
20200921_1828.cr2  Exif.Photo.LensSpecification                 Rational    4  18/1 35/1 0/1 0/1
20200921_1828.cr2  Exif.Photo.LensModel                         Ascii       8  18-35mm
20200921_1828.cr2  Exif.Photo.LensSerialNumber                  Ascii      11  0000000000
20201017_2113.cr2  Exif.CanonCs.LensType                        Short       1  Sigma 14-24mm f/2.8 DG HSM | A
20201017_2113.cr2  Exif.CanonCs.Lens                            Short       3  18.0 - 35.0 mm
20201017_2113.cr2  Exif.Canon.LensModel                         Ascii      74  18-35mm F1.8 DC HSM | Art 013
20201017_2113.cr2  Exif.Photo.LensSpecification                 Rational    4  18/1 35/1 0/1 0/1
20201017_2113.cr2  Exif.Photo.LensModel                         Ascii      30  18-35mm F1.8 DC HSM | Art 013
20201017_2113.cr2  Exif.Photo.LensSerialNumber                  Ascii      11  0000000000
```
**Expected behavior**
Lens should be detected as `Sigma 18-35mm f/1.8 DC HSM | A` and not `Sigma 14-24mm f/2.8 DG HSM | A`.

**Desktop (please complete the following information):**
 - OS: Arch Linux

**Additional context**
I hope this information is enough so I'm not attaching the RAW files here, but if needed I can attach them.

**Comment by clanmills** [comment]
You can probably fix this for yourself in the "configuration" file ~/.exiv2:

```ini
[canon]
150=Sigma 18-35mm f/1.8 DC HSM | A
368=Sigma 14-24mm f/2.8 DG HSM | A
```

You'll find this discussed on the exiv2 man page: https://exiv2.org/manpage.html 

If this fixes it, can you close the issue.  If this doesn't fix it, please get the images to me (with Dropbox or Google Drive) and I'll investigate.

**Comment by lbschenkel** [comment]
I believe that you meant:
```ini
[canon]
150=Sigma 18-35mm f/1.8 DC HSM | A
368=Sigma 18-35mm f/1.8 DC HSM | A
```
Which works but I'm reporting this with the intent of trying to make this work out of the box for all users [1] instead of just myself.

[1] if possible: I haven't looked at how `exiv2` does it, but `exiftool` uses heuristics based on the combination of different tags

**Comment by lbschenkel** [comment]
The funny thing is that I can see code `368` assigned to this lens here: https://github.com/Exiv2/exiv2/blob/5547d10bc86564c77d0c3df0ac74dc936a9b9d45/src/canonmn_int.cpp#L1057

And it looks the change that introduced it matches my situation exactly: https://github.com/Exiv2/exiv2/commit/4f1a907792fbed08360447f5725bc1d677a040a8

Not quite sure why my `exiv2` still misdetects it, though. This is how Arch Linux builds it: https://github.com/archlinux/svntogit-packages/blob/packages/exiv2/trunk/PKGBUILD

Directions on how to proceed? Should I upload a RAW file?

**Comment by clanmills** [comment]
Lens recognition has been time sink on the engineering resources of Team Exiv2.  So, I introduced the ~/.exiv2 "Configuration File" in 0.26 to save lots of work and give users an instant way to recognised their lens.  You don't need to way on the release cycles of exiv2 and your distribution.  You get it fixed instantly.

I'm currently writing a book about Exiv2 as I would like to retire on my 70th birthday on 2021-01-18.  There will be an open source graphics conference in May 2021 in Rennes and which I'd like to talk about my book and run a work shop on _**Image Metadata and Exiv2 Architecture**_.  Assuming somebody takes over the maintenance of Exiv2, I'm going to make a proposal to "solve" the lens problem.  The proposal is discussed here:  http://clanmills.com/exiv2/book/#future

The lens problem is difficult.  The lens isn't stored in the metadata.  Different manufacturers use different ways to deal with this and it's very common that a number such as "368" is used to represent several lenses.  Then we have to examine other metadata to make a guess about which lens is being used.  My proposal deals with those possibilities and makes lens recognition a project that can be used by Exiv2, ExifTool and others.  So, the burden maintenance of "the heuristic" is removed from Exiv2 and ExifTool and put into M2LScript. 

For the moment, that's the way things are and I don't want to get involved in updating the existing lens recognition code.  I'm working on the book and retiring.  Will M2LScript come to life?  We'll see.

**Comment by lbschenkel** [comment]
Sure, I understand. As I said, this report had the intention of getting this fixed upstream so other users could benefit from a fix -- I am aware that I can fix the problem for myself via the configuration file. And I'm also aware that lens detection is a tricky matter (especially for third-party lenses). But thank you for taking the time to explain anyway.

I didn't know that you were retired (or close to retiring). I'm sad for the project but I want to thank you for maintaining `exiv2` for all this time.

Now that I found out that `exiv2` is *supposed* to be detecting this lens, but it isn't, I'll try to build it debug it locally. I might be able to find out what's wrong and hopefully submit a patch (or at least describe what's going on).

**Comment by clanmills** [comment]
It's probably really easy to fix this.  However, I like to have an image in the test suite for every lens to be sure that a future change won't undo your change.  It's all rather tedious.  Usually takes me about an hour to deal with a new lens, provided it doesn't collide with an existing lens.

The M2Lscript concept removes lens recognition from Exiv2 and ExifTool.  M2Lscripts can be updated (over the internet) as new lenses appear.

**Comment by lbschenkel** [comment]
OK, here is what I found out. The lens is correctly detected in `master`, but not the `0.27` maintenance branch. I was able to bisect this branch and the commit that broke it was this: https://github.com/Exiv2/exiv2/commit/b7bd0988d5ce0e59c5fefeb14332e49d1a4913da

The original commit that introduced support for this lens, before it was undone by the commit above, was this: https://github.com/Exiv2/exiv2/commit/cf4f14c67b71d889eab89fdd4f79b5ebd23267b9

**Comment by lbschenkel** [comment]
This is the diff against `v0.27.3` that solved it for me:
```diff
diff --git a/src/canonmn_int.cpp b/src/canonmn_int.cpp
index b3443819..75107ec5 100644
--- a/src/canonmn_int.cpp
+++ b/src/canonmn_int.cpp
@@ -918,7 +918,7 @@ namespace Exiv2 {
         {  137, "Tamron SP 60mm f/2 Macro Di II"                            }, // 12
         {  137, "Sigma 10-20mm f/3.5 EX DC HSM"                             }, // 13
         {  137, "Tamron SP 24-70mm f/2.8 Di VC USD"                         }, // 14
-        {  137, "Sigma 18-35mm f/1.8 DC HSM"                                }, // 15
+        {  137, "Sigma 18-35mm f/1.8 DC HSM | A"                            }, // 15
         {  137, "Sigma 12-24mm f/4.5-5.6 DG HSM II"                         }, // 16
         {  137, "Sigma 70-300mm f/4-5.6 DG OS"                              }, // 17
         {  138, "Canon EF 28-80mm f/2.8-4L"                                 },
@@ -1147,6 +1147,7 @@ namespace Exiv2 {
         {  368, "Sigma 105mm f/1.4 DG HSM"                                  }, // 8
         {  368, "Sigma 14-24mm f/2.8 DG HSM"                                }, // 9
         {  368, "Sigma 70mm f/2.8 DG Macro"                                 }, // 10
+        {  368, "Sigma 18-35mm f/1.8 DC HSM | A"                            }, // 11
         {  488, "Canon EF-S 15-85mm f/3.5-5.6 IS USM"                       },
         {  489, "Canon EF 70-300mm f/4-5.6L IS USM"                         },
         {  490, "Canon EF 8-15mm f/4L Fisheye USM"                          },
@@ -1281,6 +1282,7 @@ namespace Exiv2 {
         { 250, printCsLensByFocalLength }, // not tested
         { 254, printCsLensByFocalLength },
         { 255, printCsLensByFocalLength }, // not tested
+        { 368, printCsLensByFocalLength },
         { 491, printCsLensByFocalLength },
         { 493, printCsLensByFocalLength }, // not tested
         { 624, printCsLensByFocalLengthTC },
```
(The first change was just to make the lens name consistent.)

I hope this helps. I haven't run the test suite yet, so it's possible there's more to be done.

**Comment by clanmills** [comment]
I guess that I'm not permitted to say "No" to this.  Please extract the .exv file from your image and I'll update the test suite.  You extract the file with the command: `$ exiv2 -ea --force --verbose my.image`.  The command reports the location of the .exv which you should attach here.  It's just a little file of the metadata in your image.  It's typically about 10k.

I'll add this, provided it doesn't break anything in test suite on either 0.27-maintenance or master.

**Comment by lbschenkel** [comment]
I have uploaded both `.exv` files to https://drive.google.com/file/d/1XaCRuWl0l4mkAPw-7TzZq71e--DwJAmq/view?usp=sharing
Please let me know when you have fetched it, so I can remove the file (because of the embedded serial numbers in the EXIF, I prefer not to have them permanently attached here).

> I guess that I'm not permitted to say "No" to this.

I genuinely couldn't figure out if you were being ironic or not (English is not my mother tongue), so apologies if I'm appearing to be too entitled or too pushy. My intent is to try my best to do as much as the work as possible to get a fix upstream, so others with the same lens can also benefit (especially when using software such as Darktable). I don't want to create more work for you, so I can try to figure out how the test suite works and submit a (hopefully) ready-to-be-merged PR.

**Comment by clanmills** [comment]
Thanks for sending the files.  You can delete them now.  I've removed the serial numbers with the command:

```bash
$ exiv2 -M'del Exif.Canon.InternalSerialNumber' -M'del Exif.Photo.BodySerialNumber' -M'del Exif.Photo.LensSerialNumber' *.exv
```

I'm not sure what I meant by _I guess that I'm not permitted to say "No" to this._.  I got up at 7am and saw your message and I thought _This guy isn't going to leave me alone.  I'll get the .exv file and finish the job.  I don't want this to drag on when I want to work on my book_.

So, I'll deal with the test suite this evening.  I'll prepare the test code and post it here.  Perhaps you can prepare a PR with everything - your code changes and the test suite.  Please prepare one for 0.27-maintenance and for master.  If you submit the PR, you'll be rightfully acknowledged as the author of the change.

I do appreciate the effort you've put into discovering how this stuff "sort of" works.  And of course, the code is different for every manufacturer.  Lens recognition was added to Exiv2 before I joined the project in 2008.  I would never have added that.  We should only read/write metadata.  Playing "guess the lens" has consumed a lot of effort.

**Comment by clanmills** [comment]
Instead of doing the test code, let me explain how it's constructed and you can deal with that in your PR.  If you prefer to leave the test suite changes to me, let me know.  However, you seem enthusiastic to "dig into this" and I'm enthusiastic to give you encouragement.

There are numerous test involving lenses:

```bash
543 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/tests/bugfixes/github $ grep -l Lens *.py
test_issue_1014.py
test_issue_1069.py
test_issue_1078.py
test_issue_1208.py
test_issue_20.py
test_issue_227.py
test_issue_247.py
test_issue_45.py
test_issue_598.py
test_issue_743.py
test_issue_990.py
test_pr_1105.py
test_pr_317.py
test_pr_992.py
test_regression_issue_201.py
544 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/tests/bugfixes/github $ 
```

Have a look at https://github.com/Exiv2/exiv2/pull/992  The user provided a test python file tests/bugfixes/github/test_pr_992.py

```python
# -*- coding: utf-8 -*-

import system_tests

class NikonSigmaLens_APO_MACRO_180_F35_EX_DG_HSM(metaclass=system_tests.CaseMeta):
    url = "https://github.com/Exiv2/exiv2/pull/992"
    
    filename = "$data_path/Sigma_APO_MACRO_180_F3.5_EX_DG_HSM.exv"
    commands = ["$exiv2 -pa --grep lensid/i $filename"]
    stderr = [""]
    stdout = [""
        """Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma APO Macro 180mm F3.5 EX DG HSM
"""
]
    retval = [0]
```

The test file is test/data/Sigma_APO_MACRO_180_F3.5_EX_DG_HSM.exv and he executes the program `exiv2 -pa --grep lensid/i foo.exv` and compares the output to stdout.  That's it.

There are two test suites
1 test/ are written in bash 
- test/*.sh are the bash scripts
- test/data are reference files
- test/tmp is used to store script output (for comparison to reference output)

2 tests/ are written in python
- the tests are run with the command `python3 runner.py`
- tests/bugfixes/{redmine | github}/*.py have python scripts which also contain code and reference output
- tests/bash_tests have python scripts which will replace test/*.sh in the next release
- in the next release, the test suite will be python (except the C++ unit tests and C++ version test)

```bash
541 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/tests $ python3 runner.py --help
usage: runner.py [-h] [--config_file CONFIG_FILE] [--verbose] [--debug] [dir_or_file]

The system test suite

positional arguments:
  dir_or_file           root directory under which the testsuite searches for tests or asingle file which tests are run (defaults to the config
                        file'slocation)

optional arguments:
  -h, --help            show this help message and exit
  --config_file CONFIG_FILE
                        Path to the suite's configuration file
  --verbose, -v         verbosity level
  --debug               enable debugging output
542 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/tests $ 
```

Write a new script tests/bugfixes/sigma_18_35_DC_HSM.py (and chmod +x).  You should choose a class for your test that is unique.  You can debug it with Visual Studio Code (or probably other python debuggers).  Test it from the terminal with:

```
$ cd tests
$ python3 runner3.py --verbose bugfixes/github/sigma_18_35_DC_HSM.py
```

When you'r econfident that it's working, run the complete test suite.

```bash
$ cd build
$ make tests
```

The command: `$ make tests` executes the command `$ cd tests ; python3 runner.py` which recursively searches bugfixes and executes every python file.

It would be very helpful for you to verify my instructions (written from memory without testing) and I'll cut'n'paste our conclusion into the book *AND* acknowledge your contribution on page 2.

The test suite on 0.27-maintenance is considerably ahead of master at this time.  When master ships (as 0.28 in 2021), 'master' will be upgraded.  I haven't worked in master for more than two years, so it may be a little different from my description.  And of course my description could easily have errors - just to challenge you!

**Comment by lbschenkel** [comment]
The changes to the test suite were indeed straightforward. I have submitted the PR at https://github.com/Exiv2/exiv2/pull/1373.

Thank you for your patience and willingness to educate me during your retirement. I think that I will be self-sufficient if I find a similar issue in the future.

**Comment by clanmills** [comment]
Closed by #1373.  It would be helpful if you could submit another PR to master with your test/data files and tests/bugfixes code.   Please give the PR the title `Fix_1368 Sigma 18-35mm f/1.8 DC HSM (master only)`.
